ddtrace/settings/_inferred_base_service.py

```diff
@@ -0,0 +1,227 @@
+
+        for arg in args:
+            has_flag_prefix = arg.startswith("-")
+            is_env_variable = "=" in arg
+
+            should_skip_arg = prev_arg_is_flag or has_flag_prefix or is_env_variable
+
+            if module_flag:
+                return ServiceMetadata(arg), True
+
+            if not should_skip_arg:
+                abs_path = pathlib.Path(arg).resolve()
+                if not abs_path.exists():
+                    continue
+                stripped = abs_path
+                if not stripped.is_dir():
+                    stripped = stripped.parent
+                value, ok = self.deduce_package_name(stripped)
+                if ok:
+                    return ServiceMetadata(value), True
+                return ServiceMetadata(self.find_nearest_top_level(stripped)), True
+
+            if has_flag_prefix and arg == "-m":
+                module_flag = True
+
+            prev_arg_is_flag = has_flag_prefix
+
+        return ServiceMetadata(""), False
+
+    def deduce_package_name(self, fp: pathlib.Path) -> Tuple[str, bool]:
+        # Walks the file path until a `__init__.py` is not found.
+        # All the dir traversed are joined then with `.`
+        up = pathlib.Path(fp).parent
+        current = fp
+        traversed: List[str] = []
+
+        while current != up:
+            if not (current / INIT_PY).exists():
+                break
+            traversed.insert(0, current.name)
+            current = up
+            up = current.parent
+
+        return ".".join(traversed), len(traversed) > 0
+
+    def find_nearest_top_level(self, fp: pathlib.Path) -> str:
+        # returns the top level dir containing a .py file starting walking up from fp
+        up = fp.parent
+        current = fp
+        last = current
+
+        while current != up:
+            if not fnmatch.filter(os.listdir(current), ALL_PY_FILES):
+                break
+            last = current
+            current = up
+            up = current.parent
+
+        return last.name
+
+
+class GunicornDetector:
+    def __init__(self, ctx):
+        self.ctx = ctx
+        name = "gunicorn"
+        self.name = name
+        self.pattern = name
+
+    def expected_command_name(self) -> str:
+        return "gunicorn"
+
+    def detect(self, args: List[str]) -> Tuple[ServiceMetadata, bool]:
+        from_env = os.getenv(GUNICORN_CMD_ARGS)
+        if from_env:
+            name, ok = self.extract_gunicorn_name_from(from_env.split())
+            if ok:
+                return ServiceMetadata(name), True
+
+        wsgi_app = os.getenv(WSGI_APP_ENV)
+        if wsgi_app:
+            return ServiceMetadata(self.parse_name_from_wsgi_app(wsgi_app)), True
+
+        name, ok = self.extract_gunicorn_name_from(args)
+        if ok:
+            return ServiceMetadata(name), True
+
+        return ServiceMetadata("gunicorn"), True
+
+    def extract_gunicorn_name_from(self, args: List[str]) -> Tuple[str, bool]:
+        skip = False
```

Comment:
wouldn't the gunicorn integration provide a better / more reliable name?

Reply:
AFAIK we don't have an actual `gunicorn` integration. we patch `gevent` which is used by `gunicorn`, so I don't think we have any special service naming in place for gunicorn other than what's provided by the patch which is nothing specific to gunicorn. Possibly @tabgok or @mabdinur may know more here.